### PR TITLE
Port #243 to the Prost branch

### PR DESCRIPTION
**Pull request by nrc**
This is a straight port of #243, but not a cherry-pick because a few changes needed to be made to use Prost, the diff didn't apply cleanly because of earlier changes to the branch, and this branch is on the 2018 edition

PTAL @overvenus @ice1000

### CODE REVIEW

File: src/progress.rs [hunk 1]
```diff
@@ -278,25 +283,42 @@ impl Progress {
...
-            if rejected <= self.matched {
+            // Or rejected equals to matched and request_snapshot is the INVALID_INDEX.
+            if rejected < self.matched
+                || (rejected == self.matched && request_snapshot == INVALID_INDEX)
+            {
                 return false;
             }
             self.next_idx = self.matched + 1;
+            if request_snapshot == INVALID_INDEX {
+                self.next_idx = self.matched + 1;
```
Comment (ice1000) [inline src/progress.rs]:
```rust
self.next_idx = self.matched + 1;
if request_snapshot == INVALID_INDEX { 
    self.next_idx = self.matched + 1; 
```

Are you sure? I think there's a better way of writing this.

Comment (nrc) [reply src/progress.rs]: I just applied the diff and fixed it up. This is from https://github.com/pingcap/raft-rs/pull/243/files#diff-2731f87d144f80edb31dac3d81fd9107R296

Comment (ice1000) [reply src/progress.rs]: This line was deleted https://github.com/pingcap/raft-rs/pull/243/files#diff-2731f87d144f80edb31dac3d81fd9107L288

So in your case you should probably delete line 288 too

Comment (nrc) [reply src/progress.rs]: Oh right, d'oh. Thanks

**Pull request merged**